Add solo statistics watcher component to deliver incremental global user statistics updates

bdach opened the pull request:
- Continuation of https://github.com/ppy/osu/issues/18877
- Needs https://github.com/ppy/osu-server-spectator/pull/157 to work

Due to the client-side design that I decided on for this flow (for not much reason other than that it seemed better than the alternative _somehow_) this component is the main workhorse of delivering the updates. The main point is that the watcher subscribes to the events that spectator server sends out on every successful processing of a score, issues API requests to get the user's global stats, and therefore can issue incremental updates to whichever game component wants them.

I'm aware that this is ~~dumbed-down~~ simplistic in places. As written, if an API call fails, there is no retry other than  what `APIAccess` provides. Thus, if you set a score, the API request to get stats after that fails, and then you set another score and the second API request to get stats succeeds, the next differential update issued will merge both scores and send an incremental update that contains the effects of both. The same will happen if an update isn't explicitly requested for a score (which will happen when retrying or failing).

I'm PRing this mostly to see what the response is, as I kinda can't bear to look at this diff anymore since I've been staring at it all day to try and provide all the nice guarantees I wanted out of it (full isolation - i.e. each update after a score sees only the effects of that one score - and recovery after API failures), but I don't think I'm able to reliably make that work client-side. See [relevant discord conversation](https://discord.com/channels/188630481301012481/188630652340404224/1055479527490465832). If the nice guarantees are indeed desired, I'd probably be looking to make the score stats queue send out all of that data via redis.

If wishing to test this full-stack, the following diff can be used to issue simple notifications using this component:

<details>
<summary>diff</summary>

```diff
diff --git a/osu.Game/OsuGameBase.cs b/osu.Game/OsuGameBase.cs
index 7586dc6407..36fd5a4177 100644
--- a/osu.Game/OsuGameBase.cs
+++ b/osu.Game/OsuGameBase.cs
@@ -46,6 +46,7 @@
 using osu.Game.Online.Chat;
 using osu.Game.Online.Metadata;
 using osu.Game.Online.Multiplayer;
+using osu.Game.Online.Solo;
 using osu.Game.Online.Spectator;
 using osu.Game.Overlays;
 using osu.Game.Overlays.Settings;
@@ -193,6 +194,7 @@ public virtual string Version
         protected MultiplayerClient MultiplayerClient { get; private set; }
 
         private MetadataClient metadataClient;
+        private SoloStatisticsWatcher soloStatisticsWatcher;
 
         private RealmAccess realm;
 
@@ -301,6 +303,7 @@ private void load(ReadableKeyCombinationProvider keyCombinationProvider)
             dependencies.CacheAs(spectatorClient = new OnlineSpectatorClient(endpoints));
             dependencies.CacheAs(MultiplayerClient = new OnlineMultiplayerClient(endpoints));
             dependencies.CacheAs(metadataClient = new OnlineMetadataClient(endpoints));
+            dependencies.CacheAs(soloStatisticsWatcher = new SoloStatisticsWatcher());
 
             AddInternal(new BeatmapOnlineChangeIngest(beatmapUpdater, realm, metadataClient));
 
@@ -346,6 +349,7 @@ private void load(ReadableKeyCombinationProvider keyCombinationProvider)
             AddInternal(spectatorClient);
             AddInternal(MultiplayerClient);
             AddInternal(metadataClient);
+            AddInternal(soloStatisticsWatcher);
 
             AddInternal(rulesetConfigCache);
 
diff --git a/osu.Game/Screens/Ranking/SoloResultsScreen.cs b/osu.Game/Screens/Ranking/SoloResultsScreen.cs
index 3774cf16b1..709b26b50c 100644
--- a/osu.Game/Screens/Ranking/SoloResultsScreen.cs
+++ b/osu.Game/Screens/Ranking/SoloResultsScreen.cs
@@ -10,6 +10,9 @@
 using osu.Game.Beatmaps;
 using osu.Game.Online.API;
 using osu.Game.Online.API.Requests;
+using osu.Game.Online.Solo;
+using osu.Game.Overlays;
+using osu.Game.Overlays.Notifications;
 using osu.Game.Rulesets;
 using osu.Game.Scoring;
 
@@ -22,11 +25,27 @@ public partial class SoloResultsScreen : ResultsScreen
         [Resolved]
         private RulesetStore rulesets { get; set; }
 
+        [Resolved]
+        private SoloStatisticsWatcher soloStatisticsWatcher { get; set; }
+
+        [Resolved]
+        private INotificationOverlay notifications { get; set; }
+
         public SoloResultsScreen(ScoreInfo score, bool allowRetry)
             : base(score, allowRetry)
         {
         }
 
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            soloStatisticsWatcher.RegisterForStatisticsUpdateAfter(Score, update => notifications.Post(new SimpleNotification
+            {
+                Text = $"Total score: {update.Before?.TotalScore} -> {update.After.TotalScore}\nPlaytime: {update.Before?.PlayTime} -> {update.After.PlayTime}"
+            }));
+        }
+
         protected override APIRequest FetchScores(Action<IEnumerable<ScoreInfo>> scoresCallback)
         {
             if (Score.BeatmapInfo.OnlineID <= 0 || Score.BeatmapInfo.Status <= BeatmapOnlineStatus.Pending)
```

</details>

bdach committed:
Add data structure for delivering statistics updates

bdach committed:
Implement solo statistics watcher

bdach committed:
Add test coverage for statistics watcher

bdach committed:
Refactor test to be easier to work with

bdach committed:
Limit tracking unhandled scores to just the last one

bdach committed:
Add test coverage of merging ignored score updates

bdach committed:
Ignore statistics update requests from third-party rulesets for now

bdach committed:
Fix solo statistics watcher firing requests for invalid user with id 1

Can happen during login flow (see `APIAccess.attemptConnect()`).

bdach committed:
Ensure latest stats are cleared on successful profile fetch

peppy commented on osu.Game/Online/Solo/SoloStatisticsWatcher.cs:
```diff
@@ -0,0 +1,139 @@
+                return;
+            }
+
+            callbacks[score.OnlineID] = callback;
```
Just for sanity, shall we `Debug.Assert` or `throw` if attempting to register for callback when one already exists? Right now it will silently overwrite.

bdach replied:
I can do `.Add()` which will have the same effect.

bdach committed:
Fix wrong handling of missing ruleset statistics

bdach committed:
Add callback to dictionary rather than overwrite

Attempting to overwrite will henceforth throw an exception.

bdach commented:
As I was doing full-stack testing with the last branch (for now) in this series, I noticed some funky stuff that stemmed from some invalid assumptions, and so I've pushed fixes for them to this branch. In order:

- d6e079a2b4e09cb80a30189c84a57c9eb3605f83: Should be self-explanatory.
- fd9110a61e8424384ad90fbea0def6101e3ed430: As it turns out, during the login flow, the local user will always briefly have an ID of 1, due to this logic:

  https://github.com/ppy/osu/blob/master/osu.Game/Online/API/APIAccess.cs#L192-L198

  This is now explicitly blocked to not fire bogus requests.
- 3c26016b61bd43e85c386c69e447665aece584e6: More for safety rather than anything else, as I got a "key already in dictionary" exception once when debugging with a breakpoint inside the request handler.
- 6c4ca387e01fcbbe82b1b29c49e0b0011f2cb1d3: As it turns out, web will return no key-value pair rather than a key-value pair with empty statistics if a user hasn't set a score on a given ruleset yet.
- 78c47a3695a9e8c1bbfa1ef8f07eb7ae2f3df5f4: Addresses review above.

peppy commented:
With the changes, I'm generally okay with this structure. I'm not looking ahead to subsequent changes as I think it's easier to review in isolation.
